File: __tests__/core/result.test.ts
/**
 * Unit-тесты для Result<T> (discriminated union).
 */

import {success, failure} from '../../src/core/errors/result';
import type {Result} from '../../src/core/errors/result';

describe('Result<T>', () => {
  describe('Success', () => {
    it('создаёт Success с данными', () => {
      const result = success('hello');
      expect(result.ok).toBe(true);
      if (result.ok) {
        expect(result.data).toBe('hello');
      }
    });

    it('работает с числами', () => {
      const result = success(42);
      expect(result.ok).toBe(true);
      if (result.ok) {
        expect(result.data).toBe(42);
      }
    });

    it('работает с объектами', () => {
      const data = {label: 'bottle', confidence: 0.95};
      const result = success(data);
      expect(result.ok).toBe(true);
      if (result.ok) {
        expect(result.data).toEqual(data);
      }
    });
  });

  describe('Failure', () => {
    it('создаёт Failure с кодом и сообщением', () => {
      const result = failure('OCR_NO_TEXT', 'Текст не обнаружен');
      expect(result.ok).toBe(false);
      if (!result.ok) {
        expect(result.errorCode).toBe('OCR_NO_TEXT');
        expect(result.userMessage).toBe('Текст не обнаружен');
      }
    });
  });

  describe('Pattern matching', () => {
    it('различает Success и Failure через ok', () => {
      const results: Result<string>[] = [
        success('данные'),
        failure('ERR', 'ошибка'),
      ];

      const outputs = results.map(r => (r.ok ? r.data : r.errorCode));
      expect(outputs).toEqual(['данные', 'ERR']);
    });
  });
});

File: test/core/errors/result_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:assist_eye/core/errors/result.dart';

/// Тесты для Result<T> — базового типа возврата всех модулей.
void main() {
  group('Result', () {
    group('Success', () {
      test('хранит данные и обеспечивает equality', () {
        // Arrange
        const result1 = Success<String>('hello');
        const result2 = Success<String>('hello');

        // Act & Assert
        expect(result1, equals(result2));
        expect(result1.data, 'hello');
      });

      test('различает разные данные', () {
        // Arrange
        const result1 = Success<String>('hello');
        const result2 = Success<String>('world');

        // Assert
        expect(result1, isNot(equals(result2)));
      });
    });

    group('Failure', () {
      test('хранит код ошибки и сообщение для пользователя', () {
        // Arrange
        const failure = Failure<String>(
          errorCode: 'OCR_NO_TEXT',
          userMessage: 'Текст не обнаружен',
        );

        // Assert
        expect(failure.errorCode, 'OCR_NO_TEXT');
        expect(failure.userMessage, 'Текст не обнаружен');
      });

      test('обеспечивает equality по полям', () {
        // Arrange
        const f1 = Failure<String>(
          errorCode: 'ERR',
          userMessage: 'msg',
        );
        const f2 = Failure<String>(
          errorCode: 'ERR',
          userMessage: 'msg',
        );

        // Assert
        expect(f1, equals(f2));
      });
    });

    test('Success и Failure — разные подтипы (sealed class pattern)', () {
      // Arrange
      const Result<String> success = Success('data');
      const Result<String> failure = Failure(
        errorCode: 'ERR',
        userMessage: 'msg',
      );

      // Assert — проверяем, что switch выражение работает
      final isSuccess = switch (success) {
        Success() => true,
        Failure() => false,
      };
      final isFailure = switch (failure) {
        Success() => false,
        Failure() => true,
      };

      expect(isSuccess, isTrue);
      expect(isFailure, isTrue);
    });
  });
}
